File: backend-javascript/02-backend-javascript-prototype/src/modules/person/person.controller.js
import { HTTP_STATUS } from '../../shared/constants/http.js';

const MESSAGES = {
  ITEM_NOT_FOUND: 'Person not found',
};

class Controller {
  constructor(service) {
    this.service = service;

    this.getItems = this.getItems.bind(this);
    this.getItemById = this.getItemById.bind(this);
    this.createItem = this.createItem.bind(this);
    this.updateItem = this.updateItem.bind(this);
    this.deleteItem = this.deleteItem.bind(this);
  }

  async getItems(req, res, next) {
    try {
      const result = await this.service.getItems(req.query);
      res.status(HTTP_STATUS.OK).json(result);
    } catch (error) {
      next(error);
    }
  }

  async getItemById(req, res, next) {
    try {
      const result = await this.service.getItemById(parseInt(req.params.id));
      if (!result) {
        return next({ status: HTTP_STATUS.NOT_FOUND, message: MESSAGES.ITEM_NOT_FOUND });
      }

      res.status(HTTP_STATUS.OK).json(result);
    } catch (error) {
      next(error);
    }
  }

  async createItem(req, res, next) {
    try {
      const result = await this.service.createItem(req.body);
      res.status(HTTP_STATUS.CREATED).json(result);
    } catch (error) {
      next(error);
    }
  }

  async updateItem(req, res, next) {
    try {
      const result = await this.service.updateItem(parseInt(req.params.id), req.body);
      if (!result) {
        return next({ status: HTTP_STATUS.NOT_FOUND, message: MESSAGES.ITEM_NOT_FOUND });
      }

      res.status(HTTP_STATUS.OK).json(result);
    } catch (error) {
      next(error);
    }
  }

  async deleteItem(req, res, next) {
    try {
      const result = await this.service.deleteItem(parseInt(req.params.id));
      if (!result) {
        return next({ status: HTTP_STATUS.NOT_FOUND, message: MESSAGES.ITEM_NOT_FOUND });
      }

      res.status(HTTP_STATUS.OK).json(result);
    } catch (error) {
      next(error);
    }
  }
}

export default Controller;

File: mcp/prototypes/002-api-rest/backend-springboot/src/main/java/com/ganatan/modules/person/PersonController.java
package com.ganatan.modules.person;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/persons")
public class PersonController {

    private final PersonService service;

    public PersonController(PersonService service) {
        this.service = service;
    }

    @GetMapping
    public ResponseEntity<?> getItems() {
        List<Person> persons = service.getItems();
        return ResponseEntity.ok(Map.of("data", persons));
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getItemById(@PathVariable Long id) {
        Person person = service.getItemById(id);
        if (person == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("error", "Person not found"));
        }
        return ResponseEntity.ok(Map.of("data", person));
    }

    @PostMapping
    public ResponseEntity<?> createItem(@RequestBody Person person) {
        Person created = service.createItem(person);
        return ResponseEntity.status(HttpStatus.CREATED)
            .body(Map.of("data", created));
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updateItem(@PathVariable Long id, @RequestBody Person person) {
        Person updated = service.updateItem(id, person);
        if (updated == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("error", "Person not found"));
        }
        return ResponseEntity.ok(Map.of("data", updated));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteItem(@PathVariable Long id) {
        boolean deleted = service.deleteItem(id);
        if (!deleted) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                .body(Map.of("error", "Person not found"));
        }
        return ResponseEntity.ok(Map.of("message", "Deleted"));
    }
}

